sway-core/src/type_system/type_id.rs

```diff
@@ -6,7 +6,7 @@ use crate::types::*;
 use super::*;
 
 /// A identifier to uniquely refer to our type terms
-#[derive(PartialEq, Eq, Hash, Clone, Copy)]
+#[derive(PartialEq, Eq, Hash, Clone, Copy, Default)]
```

Comment:
Won't this default impl just give the type id of 0, which is undefined behavior? Instead, could we remove the constraint on `TypeEngine` and use `::new()` instead of default there?

Reply:
Good catch!

Side note. I fixed it by doing this:

```rust
#[derive(Debug)]
pub(crate) struct ConcurrentSlab<I, T> {
    indexer: PhantomData<I>,
    inner: RwLock<Vec<T>>,
}

impl<I, T> Default for ConcurrentSlab<I, T> where T: Default {
    fn default() -> Self {
        Self { indexer: PhantomData, inner: Default::default() }
    }
}
```

and one would think that Rust would use `PhantomData` for `#[derive(Default)]` on `PhantomData<T>` but apparently not. It is a known issue: https://github.com/rust-lang/rust/issues/69037.